pkg/cmd/openshift-tests/disruption/poll-service/service_controller.go

```diff
@@ -0,0 +1,190 @@
+) *PollServiceController {
+
+	c := &PollServiceController{
+		nodeName:          nodeName,
+		namespaceName:     namespaceName,
+		clusterIP:         clusterIP,
+		port:              port,
+		recorder:          recorder,
+		stopConfigMapName: stopConfigMapName,
+		outFile:           outFile,
+
+		configmapLister: configmapInformer.Lister(),
+		informersToSync: []cache.InformerSynced{
+			configmapInformer.Informer().HasSynced,
+		},
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ServicePoller"),
+	}
+
+	c.syncHandler = c.syncServicePoller
+
+	configmapInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.queue.Add("checK")
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.queue.Add("checK")
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.queue.Add("checK")
+		},
+	})
+
+	return c
+}
+
+func (c *PollServiceController) syncServicePoller(ctx context.Context, key string) error {
+	_, err := c.configmapLister.ConfigMaps(c.namespaceName).Get(c.stopConfigMapName)
+	switch {
+	case err == nil:
+		c.removeAllWatchers()
+		return nil
+	case apierrors.IsNotFound(err):
+		// did not find the stopConfigMap
+	case err != nil:
+		return err
+	}
+
+	c.watcherLock.Lock()
+	defer c.watcherLock.Unlock()
+
+	if c.watcher == nil {
+		url := fmt.Sprintf("http://%s", net.JoinHostPort(c.clusterIP, c.port))
+		fmt.Fprintf(c.outFile, "Adding and starting: %v on node/%v\n", url, c.nodeName)
+
+		// the interval locator is unique for every tuple of poller to target, but the backend is per connection type
+		historicalBackendDisruptionDataForNewConnectionsName := fmt.Sprintf("%s-%v-connections", "service", monitorapi.NewConnectionType)
+		historicalBackendDisruptionDataForReusedConnectionsName := fmt.Sprintf("%s-%v-connections", "service", monitorapi.ReusedConnectionType)
+		intervalLocator := fmt.Sprintf("%s-from-node-%v-to-url-%v", "service", c.nodeName, url)
+		c.watcher = &watcher{
+			address: c.clusterIP,
+			port:    c.port,
+			newConnectionSampler: backenddisruption.NewSimpleBackendWithLocator(
+				monitorapi.NewLocator().LocateDisruptionCheck(historicalBackendDisruptionDataForNewConnectionsName, intervalLocator, monitorapi.NewConnectionType),
+				url,
+				"",
+				monitorapi.NewConnectionType,
+			),
+			reusedConnectionSampler: backenddisruption.NewSimpleBackendWithLocator(
+				monitorapi.NewLocator().LocateDisruptionCheck(historicalBackendDisruptionDataForReusedConnectionsName, intervalLocator, monitorapi.ReusedConnectionType),
+				url,
+				"",
+				monitorapi.ReusedConnectionType,
+			),
+		}
+		c.watcher.newConnectionSampler.StartEndpointMonitoring(ctx, c.recorder, nil)
+		c.watcher.reusedConnectionSampler.StartEndpointMonitoring(ctx, c.recorder, nil)
+
+		fmt.Fprintf(c.outFile, "Successfully started: %v on node/%v\n", url, c.nodeName)
+	}
+	return nil
+}
+
+func (c *PollServiceController) removeAllWatchers() {
```

Comment:
this function should call .Stop on all the samplers and set the watcher to nil, right?

Reply:
yes whoops